var/spack/repos/builtin/packages/perl/package.py

```diff
@@ -92,6 +95,52 @@ class Perl(Package):  # Perl doesn't use Autotools, it should subclass Package
 
     phases = ['configure', 'build', 'install']
 
+    @classmethod
+    def determine_spec_details(cls, prefix, exes_in_prefix):
+        exe_to_path = dict(
+            (os.path.basename(p), p) for p in exes_in_prefix
+        )
+        if 'perl' not in exe_to_path:
+            return None
+
+        perl = spack.util.executable.Executable(exe_to_path['perl'])
+        output = perl('--version', output=str)
+        version_str = ''
+        shared = False
+        threads = False
+        cpanm = False
+        if output:
+            match = re.search(r'perl.*\(v([0-9.]+)\)', output)
+            if match:
+                version_str = match.group(1)
+        else:
+            return None
+        output = perl('-V', output=str)
+        if output:
+            match = re.search(r'-Duseshrplib', output)
+            if match:
+                shared = True
+            match = re.search(r'-Duse.threads', output)
+            if match:
+                threads = True
+        if 'cpanm' in exe_to_path:
+            cpanm = True
+
+        def bool2str(inarg):
+            if inarg:
+                return "+"
+            else:
+                return "~"
+
+        if version_str:
+            spec_str = 'perl@{0}'.format(version_str)
+            spec_str += bool2str(shared) + "shared"
+            spec_str += bool2str(threads) + "threads"
+            spec_str += bool2str(cpanm) + "cpanm"
+            return Spec(spec_str)
```

Comment:
Can you change this to pack a dict above and create a Spec(perl) and set the variants on it directly?

Reply:
Redid this based on our conversation